File: go/solver.go
package main

import (
	"fmt"
	"math"
	"math/bits"
)

type sudoku struct {
	answer  board
	rows    [9]uint16
	columns [9]uint16
	regions [9]uint16
}

func solve(quiz board) (board, error) {
	s := new(sudoku)
	for i := 0; i < len(quiz); i++ {
		if quiz[i] != 0 {
			s.update(i, quiz[i])
		}
	}

	return solveInternal(s)
}

func solveInternal(s *sudoku) (board, error) {
	updated := true
	solved := false

	for updated {
		updated = false
		solved = true

		for i := 0; i < len(s.answer); i++ {
			if s.answer[i] != 0 {
				continue
			}
			solved = false

			candidate := s.getCandidateBit(i)
			if candidate == 0 {
				return s.answer, fmt.Errorf("Unresolved")
			}

			highest := highestOneBit(candidate)
			lowest := lowestOneBit(candidate)
			if highest == lowest {
				s.update(i, toNum(int(lowest)))
				updated = true
			}
		}
	}

	if solved {
		return s.answer, nil
	}

	// 候補が少ない箇所を探す
	count := math.MaxInt32
	index := 0
	for i := 0; i < len(s.answer); i++ {
		if s.answer[i] != 0 {
			continue
		}

		c := bits.OnesCount16(s.getCandidateBit(i))
		if c < count {
			count = c
			index = i
		}
	}

	// 複数ある候補のうちの一つを仮定して、再帰的に解析
	candidate := s.getCandidateBit(index)
	for i := 0; i < 9; i++ {
		bit := candidate & (1 << i)
		if bit == 0 {
			continue
		}

		newS := &sudoku{
			answer:  s.answer,
			rows:    s.rows,
			columns: s.columns,
			regions: s.regions,
		}
		newS.update(index, toNum(int(bit)))

		if answer, err := solveInternal(newS); err == nil {
			return answer, nil
		}
	}

	return s.answer, fmt.Errorf("Unresolved")
}

func (s *sudoku) update(index int, number uint8) {
	s.answer[index] = number

	r := index / 9
	c := index % 9
	value := uint16(1 << (number - 1))
	s.rows[r] |= value
	s.columns[c] |= value
	s.regions[(r/3)*3+(c/3)] |= value
}

func (s *sudoku) getCandidateBit(index int) uint16 {
	r := index / 9
	c := index % 9

	bit := uint16(0)
	bit |= s.rows[r]
	bit |= s.columns[c]
	bit |= s.regions[(r/3)*3+(c/3)]

	return ^bit & 0x1FF
}

func highestOneBit(i uint16) uint16 {
	return i & (math.MaxUint16 >> bits.LeadingZeros16(i))
}

func lowestOneBit(i uint16) uint16 {
	return i & -i
}

func toNum(bit int) uint8 {
	// 対数関数で計算できるが、パターンが少ないので switch の方が速い
	switch bit {
	case 1:
		return 1
	case 2:
		return 2
	case 4:
		return 3
	case 8:
		return 4
	case 16:
		return 5
	case 32:
		return 6
	case 64:
		return 7
	case 128:
		return 8
	case 256:
		return 9
	default:
		panic("Invalid bit.")
	}
}

File: rust/src/solver.rs
pub const N_BOARD: usize = 9 * 9;
pub type Board = [u8; N_BOARD];

pub fn solve(quiz: Board) -> Result<Board, String> {
    let mut s = Sudoku::new();
    for i in 0..quiz.len() {
        if quiz[i] != 0 {
            s.update(i, quiz[i]);
        }
    }

    return solve_internal(&mut s);
}

fn solve_internal(s: &mut Sudoku) -> Result<Board, String> {
    let mut updated = true;
    let mut solved = false;

    while updated {
        updated = false;
        solved = true;

        for i in 0..s.answer.len() {
            if s.answer[i] != 0 {
                continue;
            }
            solved = false;

            let candidate = s.get_candidate_bit(i);
            if candidate == 0 {
                return Err("Unresolved.".to_string());
            }

            let lowest = lowest_one_bit(candidate);
            if candidate == lowest {
                s.update(i, to_num(lowest));
                updated = true;
            }
        }
    }

    if solved {
        return Ok(s.answer);
    }

    let mut count = usize::MAX;
    let mut index = 0;
    for i in 0..s.answer.len() {
        if s.answer[i] != 0 {
            continue;
        }

        let c = bit_count(s.get_candidate_bit(i));
        if c < count {
            count = c;
            index = i;
        }
    }

    let candidate = s.get_candidate_bit(index);
    for i in 0..9 {
        let bit = candidate & (1 << i);
        if bit == 0 {
            continue;
        }

        let mut new_s = Sudoku {
            answer: s.answer,
            rows: s.rows,
            columns: s.columns,
            regions: s.regions,
        };
        new_s.update(index, i + 1);

        if let Ok(answer) = solve_internal(&mut new_s) {
            return Ok(answer);
        }
    }

    return Err("Unresolved.".to_string());
}

fn lowest_one_bit(i: u16) -> u16 {
    return i & (!i + 1);
}

fn to_num(bit: u16) -> u8 {
    // 対数関数で計算できるが、パターンが少ないので switch の方が速い
    return match bit {
        1 => 1,
        2 => 2,
        4 => 3,
        8 => 4,
        16 => 5,
        32 => 6,
        64 => 7,
        128 => 8,
        256 => 9,
        _ => panic!("Invalid bit."),
    };
}

fn bit_count(value: u16) -> usize {
    let mut sum = 0;
    for i in 0..9 {
        if (value & 1 << i) != 0 {
            sum += 1;
        }
    }

    return sum;
}

struct Sudoku {
    answer: Board,
    rows: [u16; 9],
    columns: [u16; 9],
    regions: [u16; 9],
}

impl Sudoku {
    fn new() -> Sudoku {
        return Sudoku {
            answer: [0; N_BOARD],
            rows: [0; 9],
            columns: [0; 9],
            regions: [0; 9],
        };
    }

    fn update(&mut self, index: usize, number: u8) {
        self.answer[index] = number;

        let r: usize = index / 9;
        let c: usize = index % 9;
        let value: u16 = 1 << (number - 1);
        self.rows[r] |= value;
        self.columns[c] |= value;
        self.regions[(r / 3) * 3 + (c / 3)] |= value;
    }

    fn get_candidate_bit(&self, index: usize) -> u16 {
        let r: usize = index / 9;
        let c: usize = index % 9;

        let mut bit: u16 = 0;
        bit |= self.rows[r];
        bit |= self.columns[c];
        bit |= self.regions[(r / 3) * 3 + (c / 3)];

        return !bit & 0x1FF;
    }
}
